python/ray/experimental/async_plasma.py

```diff
@@ -0,0 +1,235 @@
+
+        Args:
+            future (PlasmaObjectFuture): A PlasmaObjectFuture instance.
+        """
+        future.prev = self.tail
+        if self.tail is None:
+            assert self.head is None
+            self.head = future
+        else:
+            self.tail.next = future
+        self.tail = future
+        # Once done, it will be removed from the list.
+        future.add_done_callback(self.remove)
+
+    def remove(self, future: PlasmaObjectFuture):
+        """Remove an object from the linked list.
+
+        Args:
+            future (PlasmaObjectFuture): A PlasmaObjectFuture instance.
+        """
+        if self._loop.get_debug():
+            logger.debug("Removing %s from the linked list.", future)
+        if future.prev is None:
+            assert future is self.head
+            self.head = future.next
+            if self.head is None:
+                self.tail = None
+                if not self.cancelled():
+                    self.set_result(None)
+            else:
+                self.head.prev = None
+        elif future.next is None:
+            assert future is self.tail
+            self.tail = future.prev
+            if self.tail is None:
+                self.head = None
+                if not self.cancelled():
+                    self.set_result(None)
+            else:
+                self.tail.prev = None
+
+    def cancel(self, *args, **kwargs):
+        """Manually cancel all tasks assigned to this event loop."""
+        # Because remove all futures will trigger `set_result`,
+        # we cancel itself first.
+        super().cancel()
+        for future in self.traverse():
+            # All cancelled futures should have callbacks to removed itself
+            # from this linked list. However, these callbacks are scheduled in
+            # an event loop, so we could still find them in our list.
+            if not future.cancelled():
+                future.cancel()
+
+    def set_result(self, result):
+        """Complete all tasks. """
+        for future in self.traverse():
+            # All cancelled futures should have callbacks to removed itself
+            # from this linked list. However, these callbacks are scheduled in
+            # an event loop, so we could still find them in our list.
+            future.set_result(result)
+        super().set_result(result)
+
+    def traverse(self):
+        """Traverse this linked list.
+
+        Yields:
+            PlasmaObjectFuture: PlasmaObjectFuture instances.
+        """
+        current = self.head
+        while current is not None:
+            yield current
+            current = current.next
+
+
+class PlasmaEventHandler:
+    """This class is an event handler for Plasma."""
+
+    def __init__(self, loop, worker):
+        super().__init__()
+        self._loop: asyncio.BaseEventLoop = loop
+        self._worker = worker
+        self._waiting_dict: Dict[ray.ObjectID, PlasmaObjectLinkedList] = {}
```

Comment:
Can you avoid using 3.5 syntax here. Also, can this be a normal list?

Reply:
No. Using a normal list will increase the complexity of the algorithm from O(n) to O(n^2).

Reply:
What about using a set then?

Reply:
But I want the futures in the linked list be schedule as the order they are registered. So I need an ordered set (python does not have one). In this case, I preferred linked list because it is faster.